### DISCUSSION TITLE
Usage of generics with httpsCallable functions

### DISCUSSION THREAD

**Discussion by nthonymiller** [firebase/firebase-js-sdk#4415] [Q&A] [answered]
From the documentation the callable functions as per below:

```typescript
const functions = getFunctions(firebaseApp);
const addMessage = httpsCallable(functions, 'addMessage');
addMessage({ text: messageText })
  .then((result) => ....)
```

Usage of generics would help with the function signature:
```typescript

interface AddMessageRequest {
   text: string;
}

const functions = getFunctions(firebaseApp);
const addMessage = httpsCallable<AddMessageRequest>(functions, 'addMessage');
addMessage({ text: messageText })
  .then((result) => {

```

The response could also typed as well.

**Comment by Feiyang1** [comment, marked as answer]
Thanks for the feedback! It seems to be the right thing to do. @hsubox76 What do you think?

**Reply by mbleigh** [reply]
👍  Seems like it should take both a request and response type (with each defaulting to any).

**Reply by hsubox76** [reply]
Should the generic be just a whole function type like
```
interface AddMessageResult {
  status: 'error' | 'success';
}

type AddMessage = (text: string) => Promise<AddMessageResult>;

const addMessage = httpsCallable<AddMessage>(functions, 'addMessage');
```

Or is it better to have 2 generics like

```
const addMessage = httpsCallable<AddMessageRequest, Result>
```
?

**Reply by mbleigh** [reply]
IIRC the response has a mandatory `{data: {}}` wrapper around it, so I think a request/response two-type generic is most appropriate. That way the actual response signature can be `Promise<{data: T}>`

**Reply by hsubox76** [reply]
PR: https://github.com/firebase/firebase-js-sdk/pull/4466

**Reply by clytras** [reply]
Why I can't seem to find this PR changes any more?
Is there any refactoring that removed the `httpsCallable` generic typings from the repo?

**Reply by hsubox76** [reply]
No? The PR was made before we launched v9 and made the modular API the default packages. Basically, all the directories that end with "-exp" are now just the same name without the "-exp". For example, the file api.ts from that PR is here: https://github.com/firebase/firebase-js-sdk/blob/master/packages/functions/src/api.ts

If you're seeing a problem with the types, please file an issue, those will get a faster response.

If you just wanted to see the typings for reference, our reference docs are probably the best place to go: https://firebase.google.com/docs/reference/js/functions.md#httpscallable

**Reply by clytras** [reply]
Thank you for the response @hsubox76.

I was looking for the `httpsCallable` for another Firebase SDK library for React Native; on a related issue there about `httpsCallable` types implementation, there was a claim that the Firebase JS SDK `httpsCallable` does not support generic typings.

Now it's clear to me that it does support generic types, and I can discuss with the maintainers the implementation of the `httpsCallable` typings on that particular library.
